File: radar_sdk/examples/py/BGT60ATR24C/presence_detection.py
# ===========================================================================
# Copyright (C) 2021-2022 Infineon Technologies AG
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. Neither the name of the copyright holder nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
# ===========================================================================

import argparse
from collections import namedtuple
from scipy import signal

from ifxradarsdk import get_version_full
from ifxradarsdk.fmcw import DeviceFmcw
from ifxradarsdk.fmcw.types import FmcwSimpleSequenceConfig, FmcwSequenceChirp
from helpers.fft_spectrum import *


class PresenceAntiPeekingAlgo:
    def __init__(self, num_samples_per_chirp, num_chirps_per_frame):
        """Presence and Anti-Peeking Algorithm
        
        This is a simple use case of an a presence detection and
        anti-peeking demo.
        
        Parameters:
        num_samples_per_chirp: Number of samples per chirp
        """
        self.num_samples_per_chirp = num_samples_per_chirp
        self.num_chirps_per_frame = num_chirps_per_frame

        # Algorithm Parameters
        self.detect_start_sample = num_samples_per_chirp // 8
        self.detect_end_sample = num_samples_per_chirp // 2
        self.peek_start_sample = num_samples_per_chirp // 2
        self.peek_end_sample = num_samples_per_chirp

        self.threshold_presence = 0.0007
        self.threshold_peeking = 0.0006

        self.alpha_slow = 0.001
        self.alpha_med = 0.05
        self.alpha_fast = 0.6

        # Initialize state
        self.presence_status = False
        self.peeking_status = False
        self.first_run = True

        # Use Blackmann-Harris as window function
        self.window = signal.blackmanharris(num_samples_per_chirp).reshape(1, num_samples_per_chirp)

    def presence(self, mat):
        """Run the presence and anti-peeking algorithm on the current frame.
        
        Parameters:
            - mat: Radar data for one antenna as returned by Frame.get_mat_from_antenna
        
        Returns:
            - Tuple consisting of the state for presence detection and peeking.
              The first bool indicates if a target was detected. The second bool
              indicates if peeking was detected.
        """
        # copy values into local variables to keep names short
        alpha_slow = self.alpha_slow
        alpha_med = self.alpha_med
        alpha_fast = self.alpha_fast

        # Compute range FFT
        range_fft = fft_spectrum(mat, self.window)

        # Average absolute FFT values over number of chirps
        fft_spec_abs = abs(range_fft)
        fft_norm = np.divide(fft_spec_abs.sum(axis=0), self.num_chirps_per_frame)

        # Presence sensing
        if self.first_run:  # initialize averages
            self.slow_avg = fft_norm
            self.fast_avg = fft_norm
            self.slow_peek_avg = fft_norm
            self.first_run = False

        if self.presence_status == False:
            alpha_used = alpha_med
        else:
            alpha_used = alpha_slow

        self.slow_avg = self.slow_avg * (1 - alpha_used) + fft_norm * alpha_used
        self.fast_avg = self.fast_avg * (1 - alpha_fast) + fft_norm * alpha_fast
        data = self.fast_avg - self.slow_avg

        self.presence_status = np.max(data[self.detect_start_sample:self.detect_end_sample]) > self.threshold_presence

        # Peeking sensing
        if self.peeking_status == False:
            alpha_used = self.alpha_med
        else:
            alpha_used = self.alpha_slow

        self.slow_peek_avg = self.slow_peek_avg * (1 - alpha_used) + fft_norm * alpha_used
        data_peek = self.fast_avg - self.slow_peek_avg

        self.peeking_status = np.max(data_peek[self.peek_start_sample:self.peek_end_sample]) > self.threshold_peeking

        return namedtuple("state", ["presence", "peeking"])(self.presence_status, self.peeking_status)


if __name__ == "__main__":
    # -----------------------------------------------------------
    # Arguments
    # -----------------------------------------------------------
    parser = argparse.ArgumentParser(
        description='''Derives presence and peeking information from Radar Data''')

    parser.add_argument('-n', '--nframes', type=int, default=100, help="number of frames, default 100")
    parser.add_argument('-f', '--frate', type=int, default=5, help="frame rate in Hz, default 5")

    args = parser.parse_args()

    config = FmcwSimpleSequenceConfig(
        frame_repetition_time_s=1 / args.frate,  # Frame repetition time default 0.2s (frame rate of 5Hz)
        chirp_repetition_time_s=0.000411238,  # Chirp repetition time (or pulse repetition time) of 411.238us
        num_chirps=32,  # 32 chirps per frame
        tdm_mimo=False,  # MIMO disabled
        chirp=FmcwSequenceChirp(
            start_frequency_Hz=59_133_931_281,  # start frequency: 59.133931281 GHz
            end_frequency_Hz=62_366_068_720,  # end frequency: 62.366068720 GHz
            sample_rate_Hz=1e6,  # ADC sample rate of 1MHz
            num_samples=64,  # 64 samples per chirp
            rx_mask=1,  # RX antenna 1 activated
            tx_mask=1,  # TX antenna 1 activated
            tx_power_level=31,  # TX power level of 31
            lp_cutoff_Hz=500000,  # Anti-aliasing cutoff frequency of 500kHz
            hp_cutoff_Hz=80000,  # 80kHz cutoff frequency for high-pass filter
            if_gain_dB=33,  # 33dB if gain
        )
    )

    with DeviceFmcw() as device:
        print(f"Radar SDK Version: {get_version_full()}")
        print("Sensor: " + str(device.get_sensor_type()))

        # set device config for presence sensing
        sequence = device.create_simple_sequence(config)
        device.set_acquisition_sequence(sequence)

        # initialize the algorithm
        algo = PresenceAntiPeekingAlgo(config.chirp.num_samples, config.num_chirps)

        for frame_number in range(args.nframes):
            # Get radar data for the first RX antenna
            frame_contents = device.get_next_frame()
            frame = frame_contents[0]

            # matrix of dimension num_chirps_per_frame x num_samples_per_chirp for RX1
            mat = frame[0, :, :]
            presence_status, peeking_status = algo.presence(mat)

            print(f" Presence: {presence_status}")
            print(f"       Peeking: {peeking_status}")

File: radar_sdk/examples/matlab/BGT60TR13C/presence_detection.m
% ===========================================================================
% Copyright (C) 2021-2022 Infineon Technologies AG
%
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are met:
%
% 1. Redistributions of source code must retain the above copyright notice,
%    this list of conditions and the following disclaimer.
% 2. Redistributions in binary form must reproduce the above copyright
%    notice, this list of conditions and the following disclaimer in the
%    documentation and/or other materials provided with the distribution.
% 3. Neither the name of the copyright holder nor the names of its
%    contributors may be used to endorse or promote products derived from
%    this software without specific prior written permission.
%
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
% AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
% IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
% ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
% LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
% CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
% SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
% INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
% CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
% ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
% POSSIBILITY OF SUCH DAMAGE.
% ===========================================================================

clear all;
addpath('../../../sdk/matlab/RadarSDKMEXWrapper');
% load package if this is not matlab
if((exist ("OCTAVE_VERSION", "builtin") > 0))
    pkg load communications
end

disp(['Radar SDK Version: ' RadarDevice.get_version_full()]);

%##########################################################################
% STEP 1: Create one DeviceConfig instance. Please refer to DeviceConfig.m
% for more details about TR13C or ATR24C (MIMO) configuration.
% Example configurations are also provided.
%##########################################################################

% Set device configuration for presence sensing
f_start = 59.0000e9; %FMCW chirp start frequency in Hz
f_end   = 61.9979e9; %FMCW chirp end frequency in Hz
samples_per_chirp  = 64;
chirps_per_frame   = 32;

oDevConf = DeviceConfig(1e6, ... % sample_rate_Hz
                       1, ... % rx_mask
                       1, ... % tx_mask
                       31, ... % tx_power_level
                       33, ... % if_gain_dB
                       f_start, ... % start_frequency_Hz
                       f_end, ... % end_frequency_Hz
                       samples_per_chirp, ... % num_samples_per_chirp
                       chirps_per_frame, ... % num_chirps_per_frame
                       0.000411238, ... % chirp_repetition_time_s
                       0.125, ... % frame_repetition_time_s
                       80000, ... % hp_cutoff_Hz
                       500000, ... % aaf_cutoff_Hz
                       0); % mimo_mode

%##########################################################################
% STEP 2: Create a RadarDevice object and connect to the Radar
%##########################################################################
Dev = RadarDevice();

%##########################################################################
% STEP 3: Configure the radar device with the required config values
%##########################################################################
Dev.set_config(oDevConf);

%##########################################################################
% STEP 4: Algorithm Configurations
%##########################################################################

zero_padding_factor = 2;

%% Range calculation
range_fft_size = (2^zero_padding_factor) * samples_per_chirp;     % zero padding factor of 2 is included
c0 = physconst('LightSpeed');                   % speed of light
BW = f_end - f_start;                           % configured bandwidth
R_max = samples_per_chirp * c0 / (2 * BW);  % maximum range
DistPerBin = R_max / range_fft_size;            % distance per frequency bin
range_len = range_fft_size / 2; % only half of the spectrum has information since the Radar data is real only
range_arr = (0:range_len-1) * DistPerBin;       % vector of distance points corresponding to frequency bins

%% Object Presence sensing algorithm parameters
range_min = 0.2;               		% minimum range of 20cm
range_max = 0.8;               		% maximum range if 80cm
threshold_presence = 0.0007;                        % threshold for object detection
alpha_slow = 0.001;                                 % slow average update coefficient
alpha_med = 0.05;                                   % medium average update coefficient
alpha_fast = 0.6;                                   % fast average update coefficient
presence_status = 0;

% Find range FFT bins for minimum and maximum ranges
temp = find(range_arr >= range_min);
detect_start_bin = temp(1);
temp = find(range_arr >= range_max);
detect_end_bin = temp(1);

%##########################################################################
% STEP 5: Fetch Radar Data
%##########################################################################

% Fetch frame data cube and reshape in the form of an antenna matrix (rows = chirps, columns = samples)
fdata = Dev.get_next_frame();
% Data matrix for the first configured receive antenna
mat = squeeze(fdata(1,:,:));

% Check shape of frame data for one antenna
if(~isequal(size(mat), [chirps_per_frame, samples_per_chirp]))
    clear Dev;
    error('Unexpected data size');
end

zero_padded = zeros(size(mat,1),size(mat,2)*2^(zero_padding_factor));
result = zeros(1,200);

hFig = figure('numbertitle', 'off');
% In this example, show frames till figure is closed.
frame_number = 0;

countdown = 0;
old_status = 0;
%##########################################################################
% STEP 6: Continue fetching Radar data and run desired algorithm on it.
%##########################################################################
while ishandle(hFig)
	%% Fetch data cube from the Radar and format it to an antenna matrix
    fdata = Dev.get_next_frame();
    if (isempty(fdata))
        continue;
    end
    % fetching frame data for the first configured antenna
    mat = squeeze(fdata(1,:,:));
    %% Run algorithm
    % compute window for applying to signal before FFT
    w = window(@blackmanharris,samples_per_chirp);

    % Mean removal and windowing of raw data
    avg_windowed = (mat - mean(mat,2)).*w';

    zero_padded(:,1:samples_per_chirp) = avg_windowed;

    % Range FFT
    range_fft = (fft(zero_padded'))/samples_per_chirp; % normalized FFT
    range_fft = 2*range_fft.'; % compensating for the negative half the spectrum
    % remove the redundant negative spectrum since FFT input was real
    range_fft = range_fft(:,1:(size(range_fft,2)/2));

    % Averaging the FFT values over number of chirps
    fft_avg = mean(range_fft);
    fft_abs = abs(fft_avg);

    if frame_number == 0  % Initialize average updates
        slow_avg = fft_abs;
        fast_avg = fft_abs;
    end

    % update the background slowly in case object is detected
    if ~presence_status
        alpha_used = alpha_med;
    else
        alpha_used = alpha_slow;
    end
    slow_avg = slow_avg*(1-alpha_used) + fft_abs*alpha_used;
    fast_avg = fast_avg*(1-alpha_fast) + fft_abs*alpha_fast;
    data = abs(fast_avg-slow_avg);

    % Detection
    detect_range_max = max(data(detect_start_bin:detect_end_bin));
    presence_status =  (detect_range_max > threshold_presence) && (detect_range_max >= max(data)-threshold_presence);

    if(presence_status == 1)
        countdown = 10;
    end

    if(old_status == 1 && presence_status==0)
        if(countdown>0)
            countdown = countdown-1;
        else
            old_status = 0;
        end
    end
    presence_output = 0;
    if(countdown > 0 )
        old_status = 1;
        presence_output = 1;
    else
        old_status = 0;
    end


    %% Update and display result
    result = [result(2:end) presence_output];
    plot(result, LineWidth=4);
    xlabel('Time',FontSize=20);
    ylabel('Detection Status', FontSize=20);
    set(gca,'xtick',[],'ytick',[]);
    set(gcf, 'name', 'Range window object sensing');
    axis([1 length(result) -0.2 1.2 ]);
    frame_number = frame_number+1;
    if(old_status)
        title(sprintf('Object detected'),FontSize=48);
    else
        title(sprintf('Absent'), FontSize=48);
    end
    drawnow
end

%##########################################################################
% STEP 6: Clear the RadarDevice object. It also stops Radar data and
% disconnects the device from the handle.
%##########################################################################
clear Dev
